File: src/main/python/com/hazza/LeetCodeByPython/SegmentOne/No14_Longest_Common_Prefix.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/4/11 21:02
# https://leetcode.com/problems/longest-common-prefix/
from typing import List


class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if not strs:
            return ''

        for i, grouped in enumerate(zip(*strs)):
            if len(set(grouped)) > 1:
                return strs[0][:i]

        return min(strs)


if __name__ == '__main__':
    solution = Solution()
    assert solution.longestCommonPrefix(["flower","flow","flight"]) == 'fl'
    assert solution.longestCommonPrefix(["dog","racecar","car"]) == ''

File: src/main/java/com/hazza/leetcode/SegmentOne/No14LongestCommonPrefix.java
package com.hazza.leetcode.SegmentOne;

/**
 * https://leetcode.com/problems/longest-common-prefix/#/description
 */
public class No14LongestCommonPrefix {
    public String longestCommonPrefix(String[] strs) {
        if(strs == null || strs.length == 0)    return "";
        String pre = strs[0];

        for(int i = 1; i < strs.length; ++i) {
            int j = 0;
            for(; j < pre.length(); ++j) {
                if (j >= strs[i].length() || pre.charAt(j) != strs[i].charAt(j))
                    break;
            }
            pre = pre.substring(0, j);
        }

        return pre;
    }
}
